### ISSUE TITLE
Add support for Git LFS

### ISSUE THREAD

**Issue by martinvonz** [martinvonz/jj#80]
Git LFS seems to be used frequently enough that it may be worth adding support for it. I don't think it'll be a priority for me very soon, but I guess that depends on how many people want it.

[The specification](https://github.com/git-lfs/git-lfs/blob/main/docs/spec.md) says that it uses clean/smudge filters. We don't have anything like that yet. So the first step is probably to add support for that. We could make the filters only available internally (i.e. in Rust code) to start with to keep it simple. On the other hand, it might not be hard to make them user-configurable.

Another option is to add a separate file type in the data model for LFS entries. For reference, we currently have [files, symlinks, trees, conflicts, and gitmodules](https://github.com/martinvonz/jj/blob/f9298582a78a4d06e6b9c5b51c67e034e7406739/lib/src/backend.rs#L291). I haven't thought through the consequences yet. There should be no difference to the user and no difference in the representation when using the Git backend. However, clean/smudge filters are probably useful to have anyway. Oh, one *possible* advantage of representing LFS entries in the model is that we can decide to always leave merged LFS files as conflicts, without downloading the files until the user checks them out or looks at the diff etc.

I don't yet know what other aspects of Git LFS we need to consider.

Originally requested in #77.

**Comment by ilyagr** [comment]
Filters are supported in libgit2, but not in rust's git2. It's also apparently not very hard to implement the logic ourselves. See https://github.com/rust-lang/git2-rs/issues/442.

**Comment by martinvonz** [comment]
I'm worried about clean/smudge in general because it seems expensive to make it behave consistently. https://git-scm.com/book/en/v2/Customizing-Git-Git-Attributes says that smudging happens just before checkout and cleaning happens just before staging. If that's correct, then that seems to mean that `git diff` on a commit that changes an LFS file would just show the changed metadata, which is not very user-friendly. On the other hand, if it's not correct, and smudging happens whenever you need to present the file to the user, then there are ugly corner cases to deal with, like where `.gitattributes` changes but subtrees remain the same (then you'd technically need to diff the whole tree recursively to tell if it actually changed). I *think* I asked someone on the Git team at Google about this and they said Git just ignores the corner case(s).

**Comment by ilyagr** [comment]
>  there are ugly corner cases to deal with, like where .gitattributes changes but subtrees remain the same

Yes, I remember setting up the LFS repo being a pain. I don't remember how git reacted to changing `.gitattributes`, but it took a while to get right; my intention is to never change the setup (which directories LFS is used for).

For reference, the setup looks like this:

```
$ cat .gitattributes
.local/bin/* filter=lfs diff=lfs merge=lfs -text
```

(I use Github's LFS support to sync a few binaries across my machines. I use `stow` to symplink to the git repo's .local/bin from my real ~/.local/bin)

My sense is that only the `filter=lfs -text` part is crucial, while `diff=lfs merge=lfs` either do nothing or are there to address the other problem you described:

> If that's correct, then that seems to mean that git diff on a commit that changes an LFS file would just show the changed metadata, which is not very user-friendly.

I think that's what the `diff` and `merge` gitattributes are for, but they don't seem to have much of an effect now:

```
$ git diff -r HEAD^
diff --git a/.local/bin/hwatch b/.local/bin/hwatch
index acc9c52..5e4ed26 100755
--- a/.local/bin/hwatch
+++ b/.local/bin/hwatch
@@ -1,3 +1,3 @@
 version https://git-lfs.github.com/spec/v1
-oid sha256:c0e1ce1ee1a4841f2df5b7721bbeefe7d91f12bfe77b0922c9ac708f93379f2b
-size 7989912
+oid sha256:4dbaaf94bfd0812a38e5eb9bc01ec0f22f0953efc13be768aa310deb6b5982ce
+size 7912656
```

This works pretty well for LFS specifically. I'm not sure what else, if anything, clean/smudge filters are used for.

This does make LFS a little different from the way `jj` treats conflicted files.

**Comment by ilyagr** [comment]
This might all be quite awkward with `jj`'s auto-rebasing. OTOH, the level of awkwardness would be similar to what happens if one tracks a binary file in a jj repository normally (without LFS), which is something we should probably eventually improve if we can (I'm not sure how).

**Comment by Ralith** [comment]
Note that LFS is widely disliked due to half-baked support from GitHub (small quotas, made worse by being consumed by [both third-party forks and CI activity](https://docs.github.com/en/repositories/working-with-files/managing-large-files/about-storage-and-bandwidth-usage)) and even less support elsewhere. I hope jj can offer a first-class solution to large binary file handling eventually, and that any LFS compat is forwards-compatible with that.

> git diff on a commit that changes an LFS file would just show the changed metadata, which is not very user-friendly

What else might they display? An actually diff of a large binary file will rarely be intelligible, even with a suitable diff algorithm (e.g. based on rolling hashes). I suppose it would be cool for a Sufficiently Smart diff viewer to be able to e.g. display two versions of a .png, but in general large binary files are opaque.

**Comment by martinvonz** [comment]
> > git diff on a commit that changes an LFS file would just show the changed metadata, which is not very user-friendly
> 
> What else might they display? An actually diff of a large binary file will rarely be intelligible

Fair enough :) But the point still stands for clean/smudge in general.

**Comment by ilyagr** [comment]
As I mentioned, there seemed to be some attempt to support intelligent diffing with LFS. I would guess that the original idea was for the user to configure custom diff tools (e.g. for pngs) that would be called by `git-lfs diff` or something like that.

**Comment by WaylonBrown** [comment]
> Note that LFS is widely disliked

Regardless of its merits, I imagine there's plenty of users like me that would love to use jj in an existing repo that uses git lfs. Whether I use jj should be opaque to others in the repo, so it wouldn't be a good reason to migrate away (to whatever solution may be better).

**Comment by martinvonz** [comment]
I agree, support for LFS would be mostly (maybe only) to make it easier to use jj with existing git repos.

**Comment by WaylonBrown** [comment]
Is there a way to ignore LFS files so that LFS-enabled repos can still use jj, even if it means we can't interact with LFS itself while using jj? Just wondering if there's a way to unblock without needing full support for LFS.

**Comment by martinvonz** [comment]
> Is there a way to ignore LFS files so that LFS-enabled repos can still use jj, even if it means we can't interact with LFS itself while using jj? Just wondering if there's a way to use jj without needing full support for LFS.

If you don't need the LFS files, then it probably already works - you'd just see the placeholder files (pointers to the real content) in the working copy, I think. But I suspect that you need the actual files, and for that I can't think of a good solution.

Oh, using sparse checkouts in a colocated repo might work. However, sparse checkouts don't currently support negative patterns, so it could be very annoying to maintain the sparse patterns depending on your repo. Hmm, it also looks like we don't have any documentation about sparse checkouts, other than `jj help sparse`. Run `jj sparse set --clear --add <path prefix> --add <path prefix> ...`. If you realize it's unmaintainable, run `jj sparse set --reset` to include all paths in the working copy again.

**Comment by waylon-brown** [comment]
So `sparse` seems like an inverse `.gitignore` (in that it's a list of path inclusions only)? Unfortunately like you said, this would probably only work for a sizable repo if it supported negative patterns, if I have it correct that `--remove` only removes an existing inclusion, rather than actually adds a negative matching pattern.

**Comment by martinvonz** [comment]
We do want to add support for arbitrary patterns in `jj sparse`. If you have time to spare, I think a good start would be to add a new `GlobMatcher` in https://github.com/martinvonz/jj/blob/main/lib/src/matchers.rs.

Then we'd also need to figure out the UX for adding and removing patterns. Git seems to use the same format as for `.gitignores` (https://git-scm.com/docs/git-sparse-checkout). It seems that you can add paths to the list with `git sparse-checkout add <pattern>`, but I didn't find a command to modify the list. Maybe you need to manually edit `.git/info/sparse-checkout` for that. Also, Git has something called "cone mode". I hope we can avoid exposing something  like that to the user.

**Comment by 71** [comment]
Instead of patterns being "prefixes" as they are now, can't we allow arbitrary globs in `jj sparse set (--add|--remove)` in terms of UI? Then you could add, remove and list patterns (rather than paths) from the CLI.

Reading comments in this issue, it seems like `jj sparse` could make it easier to work with Git LFS _and_ replace `git update-index --skip-worktree` (which I've been using a lot recently), so I could also chip in and try to help bring this.

**Comment by martinvonz** [comment]
> Instead of patterns being "prefixes" as they are now, can't we allow arbitrary globs in `jj sparse set (--add|--remove)` in terms of UI? Then you could add, remove and list patterns (rather than paths) from the CLI.

If we accept both things like `docs/` and `**/Cargo.toml`, then the issue becomes how to tell which is which. Is `docs` a file called that or should it match recursively? We could "solve" that by saying that globs are also recursive,  so the *glob* `docs` also matches all files anywhere under that directory, but I think that will make e.g. `**/Cargo.toml` confusing, because you would probably not expect that to match `lib/Cargo.toml/foo`. That's probably not much of an issue in practice for sample path (no one creates a `Cargo.toml` directory), but there are probably other examples that do happen in practice. Mercurial solves the problem by allowing [a prefix to specify what kind of pattern it is](https://www.mercurial-scm.org/doc/hg.1.html#file-name-patterns). We can just copy that solution.

> Reading comments in this issue, it seems like `jj sparse` could make it easier to work with Git LFS _and_ replace `git update-index --skip-worktree` (which I've been using a lot recently), so I could also chip in and try to help bring this.

That would be appreciated, thanks! Just to be clear, `jj sparse` makes jj completely ignore the non-sparse paths, so you will have to rely on git to populate and update those paths.

FYI, the typical use case for sparse checkouts is when you're working on only a small part of a large repo, like working only on a particular file system in the Linux repo (I have never worked in the Linux repo, so I have no idea if that's a realistic example - maybe you need most of the repo in order to do a build anyway, for example).

> **Edit**: adding/removing globs in the command line is (as the git docs mention) error-prone. I personally think this is fine as long as we add warnings about it. An alternative would be a `jj sparse edit` command which brings up `$EDITOR` on a temporary file with one pattern per line. After saving the file, `jj` parses the saved file as patterns and saves them wherever/however it wants.

Yes, I think that would be useful. I was looking for `git sparse-checkout edit` when I was typing my previous message :)

> **Edit 2**: the git docs mention that having "non-cone" globs can slow down commands, but (naively) this seems like it could be solved by compiling patterns similarly to what [`globset` does](https://github.com/BurntSushi/ripgrep/tree/master/crates/globset#performance).

There are still cases that can't be made fast, like `**/Cargo.toml`, for example. We need to visit every file in the repo to see if it matches. We can add a warning if the user adds a pattern like that. There are less clear cases like `some/dir/**/Cargo.toml`, which may be cheap if `some/dir/` is small. So maybe what we want to do is to apply the pattern, then check how many paths we visit and how many paths match, and warn if < 1% match or something. But that might be going too far :) So maybe we warn exactly when a pattern is not a pure prefix pattern (i.e. exactly what git's cone mode allows).

**Comment by elasticdog** [comment]
For posterity, there are other legit uses of Git's clean/smudge filters beyond LFS: https://github.com/elasticdog/transcrypt

**Comment by 71** [comment]
Oh, I didn't realize that `jj` erases all files not in `jj sparse list`; I thought it would simply ignore them (the way Git ignores files with `--skip-worktree`). Then I'm not sure `jj sparse` is the way to go (for my particular use case).

With that said, I actually implemented `jj sparse set --edit` _before_ I realized this, so I can submit a PR for it (and even if it's not submitted, it can serve as future reference).

**Comment by ilyagr** [comment]
> Oh, I didn't realize that jj erases all files not in jj sparse list; I thought it would simply ignore them

This confused me. My impression was the opposite: jj is only allowed to touch (or erase) files **in** `jj sparse list`, and should ignore files **not in** `jj sparse list`. Did I miss something?

**Comment by martinvonz** [comment]
> > Oh, I didn't realize that jj erases all files not in jj sparse list; I thought it would simply ignore them
> 
> This confused me. My impression was the opposite: jj is only allowed to touch (or erase) files **in** `jj sparse list`, and should ignore files **not in** `jj sparse list`. Did I miss something?

I think @71 meant when you go from having some part of the workspace populated to having that part not populated, then the `jj sparse set` command will remove those paths. For example, if you do `jj git clone <the jj repo itself>` and then `jj sparse set --clear --add src`, then all of `lib`, `docs` etc. will be removed.

@71, if you make git populate those paths after setting the sparse patterns with `jj sparse`, does that work for you?

**Comment by 71** [comment]
In my case, the problem was that I had files that I did not want in the Git repo.

1. `jj git clone ... && cd ...`
2. `echo abc > abc`
3. `jj sparse set --clear --add src`

    At this point `abc` does not exist in the working copy anymore, but `jj st` shows it. I didn't realize that the whole directory had been removed, did `jj untrack abc`, and lost the file. I can also use `jj sparse set --add abc` after 3. to recover `abc`, but am not sure how to recover `abc` through Git (without adding it back into the repo).

**Comment by Valodim** [comment]
> Regardless of LFS's merits, I imagine there's plenty of other users like me that would love to use jj in an existing repo that uses git lfs, but currently can't.

I would like to echo this sentiment: Beloved or not, LFS is fairly widely used to track binary files in repos for various reasons, and missing support for it excludes a large amount of repositories from use with jj. It's a very promising sign that "mundane" user compatibility concerns like this are taken seriously :+1:  thanks for that, and thanks for jujutsu!

**Comment by bcspragu** [comment]
This is the one thing stopping me from using `jj` everywhere. I use it for most repos, and am always forgetting the few existing LFS-using repos I can't use it with yet. Similar to other comments, I'd be fine with a workaround (e.g. ignoring files in `.gitattributes` entirely), I'd be happy to do the work and maintain the patch with a few pointers.

**Comment by martinvonz** [comment]
Thanks! That should only require changes to https://github.com/jj-vcs/jj/blob/main/lib/src/local_working_copy.rs, specifically the `snapshot()` and `check_out()` methods.

Note that this approach means that things like `jj diff` will only show the changes in the underlying files (i.e. changed hashes, I think). I hope that's okay.

**Comment by arxanas** [comment]
@bcspragu Depending on your appetite for implementation, you could also try to pick up the work I linked in this comment: https://github.com/jj-vcs/jj/issues/2920#issuecomment-2564820838. It should also handle arbitrary smudge/clean filters, submodules, etc., but needs more investment.

**Comment by bcspragu** [comment]
I have a functional (but extremely janky) patch here https://github.com/jj-vcs/jj/compare/main...bcspragu:jj:main. Some notes:

- I implemented it as a new `GitAttributesMatcher` that gets subtracted in a `DifferenceMatcher` to remove LFS files
- Modifying the matchers in `snapshot()` and `check_out()` didn't stop files mentioned in `.gitattributes` from appearing in `jj st`
  - I don't know enough of the internals of `jj` to understand why, I'm sure I'm doing something simple wrong
  - I hacked around this by updating the matcher in `cmd_status` directly. This works (and proves out the functionality), but definitely isn't correct
- Error-handling + when to actually parse the .gitattributes file needs work
- I couldn't figure out how to get a path to the repo root relative to current directory, so it currently assumes commands are run from the repo root
  - Similarly, non-root `.gitattributes` files are not currently handled
- It reuses the [`ignore` crate](https://docs.rs/ignore/latest/ignore/) for parsing the `.gitattributes`, and manually handles where the two file formats differ ([described in the Git docs](https://git-scm.com/docs/gitattributes))
  - The alternative was introducing another crate ([e.g.](https://docs.rs/git-attributes/latest/git_attributes/)), but I figured that was less desirable
- It decides a file is an LFS file if the entry in `.gitattributes` includes `filter=lfs`
  - No idea if  there's a better heuristic (or if checking `diff=...` and `merge=...` is necessary too)

> Depending on your appetite for implementation [...]

Apologies, but I'm not quite that hungry yet :sweat_smile:. Though I think that approach is a much more reasonable one in terms of what could actually be merged in

**Comment by martinvonz** [comment]
I'm not sure the `GitAttributesMatcher` solution is sufficient. There can be many `.gitattributes` sprinkled throughout the directory tree, so I think you'll need to do something similar to the `.gitignore` handling, where you create an updated matcher each time you visit a directory.

**Comment by weiznich** [comment]
@martinvonz https://github.com/weiznich/jj/blob/ignore_lfs/lib/src/matchers.rs#L592-L616 implements exactly that. I've run it locally on a repository that useses several gitattributes files in different subfolders and it seems to work fine.

**Comment by martinvonz** [comment]
Please send a PR when it's ready. That should include tests. Please put those tests in the library crate (not the CLI crate).

**Comment by bcspragu** [comment]
> @bcspragu I took some time to add support for multiple gitattributes files on top of your change in [weiznich@a125051](https://github.com/weiznich/jj/commit/a12505164ac655c8769dfee9991aae5ccf20990f).

Thanks for the updates, looks excellent!

> Please send a PR when it's ready. That should include tests. Please put those tests in the library crate (not the CLI crate).

I can pull in @weiznich's changes (if that works for them) and add some library tests. Just to confirm, the idea is to ship the LFS-ignoring functionality as a part of `jj`? Should it be behind a config option or something?

**Comment by weiznich** [comment]
@bcspragu If you have the capacity to put up a PR with tests that would be great. Just pull in my changes if that helps you.

**Comment by martinvonz** [comment]
> Just to confirm, the idea is to ship the LFS-ignoring functionality as a part of `jj`?

Yes, I'm fine with that. I assume others will object if they disagree.

> Should it be behind a config option or something?

That's probably best.

**Comment by Valodim** [comment]
Reporting in: I've been using the above mentioned branch (via `nix shell github:weiznich/jj/lfs_ignore#jujutsu`) for a brief while now, and it works very well! jj leaves lfs managed files alone, and handling them explicitly via `git lfs` commands is perfectly acceptable for my purposes.

There may be other workflows that require more direct handling of lfs stuff, but what's there (plus some polish) is a more than useful feature to ship from my PoV! 👍

**Comment by hwchen** [comment]
I know that there's a PR being worked on, but wanted to express my appreciation to @weiznich for their branch, which I'm using right now. I've only used `jj` for 2 months, and it's been quite painful for me to drop into git for just the one repo which uses LFS. And now I don't :).

**Comment by 06393993** [comment]
Below are a list of questions and design decisions I summarize from the discord chat and the PR https://github.com/jj-vcs/jj/pull/6036. I am wondering if we can come up with an agreement on all of the questions, to either pave the way to the actual implementation or more questions.
 
* Should we add a git.ignore-lfs-files config? Should it default to true or false?
    * https://github.com/jj-vcs/jj/pull/6036/files#r2036749169 suggests that we shouldn't have this config, because set it to false is never meaningful
* Should we support `.gitattributes` files only for git backend?
    * If so,
        + how to avoid `#[cfg(feature = "git")]` from scattering everywhere(see analysis below)?
        + how to hide the `.gitattributes` logic as an implementation details? When traversing the files in `local_working_copy.rs`, should we introduce a general backend state instead of exposing a `GitAttributesFile` type as opposed to the existing `GitIgnoreFile`? 
    * If not,
        + should we provide a config entry to enable or disable it? What will be the key and valid values for this config?
        + how do we enable it by default on git backend and disable it by default on other backends?
        + Are we Ok to move the gix dependency out of the `git` feature? 
        + How could the feature interact with other backends? e.g., how would the `text` attribute and `eol` attribute interact with the `.hgeol` file in fig/mercurial?
* The `GitAttributesFile` type: a hierarchical matcher to find the state of an attribute associated with a given path.
    * Should it be a public API of `jj-lib` just like `GitIgnoreFile` or not?
        + https://github.com/jj-vcs/jj/pull/6036 implements it similar to `GitIgnoreFile`.
        + This new type is likely to change dramatically as this feature won't be stable in short term.
    * How to avoid `#[cfg(feature = "git")]` from scattering everywhere?
        + https://github.com/jj-vcs/jj/pull/6036 doesn't take care of it, and should fail the compilation if the Cargo `git` feature is not enabled.
        + It is inevitable that we will add a field of this type when visiting directory entries in `local_working_copy.rs`: every directory entry can have a different result of matching if there are different `.gitattributes` files in the acestors of the path.
        + Nowadays, the `gix` dependency in `jj-lib` and `jj-cli` is behind the `git` Cargo feature.
        + The implementation will definitely use `gix` to save the effort to implement the matcher and chaining `.gitattributes` files. Therefore, the implementation is likely to behind a `#[cfg(feature = "git")]` no matter where the implementation exists: `jj-cli` or `jj-lib`.
    * Where to read the `.gitattributes` file?
        + The git documentation suggests that it will try to read both from the index(store in jj) and from the work tree(disk in jj). In the checkout process, the `.gitattributes` file in the index takes the priority, and in the checkin process, the `.gitattributes` file in the working tree takes the priority.
        + https://github.com/jj-vcs/jj/pull/6036 only reads from the disk, and probably won't work correctly if a new file under the LFS filter is added and jj is taking a snapshot.
        + What to do if the version from the store has a conflicts? pmetzger suggests that we can recover it from the last snapshot just like `.gitignore`, but I can't find related code.
        + The correct implementation should allow us to handle the `jj sparse` case out of box - the `.gitattributes` in the store will be consulted either as a priority or as a fall-back.
    * Respect the attributes set outside the `.gitattributes` file.
        + The git documentation suggests that it will also read from the following paths:
	      - $GIT_DIR/info/attributes
	      - The `core.attributesFile` git config option.
	      - If the git config is not set, $XDG_CONFIG_HOME/git/attributes or $HOME/.config/git/attributes may be used.
	      - $(prefix)/etc/gitattributes
	  + If we use `gix` correctly to create the root `GitAttributesFile` we should be fine.
* Where should we create the root `GitAttributesFile` object?
    * We have 2 options: in `local_working_copy.rs` or in `jj-cli` and pass to `TreeState` via `SnapshotOptions` or `CheckoutOptions` depending on whether we are snapshotting or updating.
    * Put everything local in `local_working_copy.rs` may result in `#[cfg(feature = "git")]` everywhere, but avoid public API changes for `jj-lib`. While the other option is vice versal.
* How does this design integrate with the EOL design?
    * It really depends on whether we want to support both features only in git backend or not.
    * If either of the features is supported outside the git backend, what's the expected behavior?
    * If both features are only supported in git backend or supported universally, they are on the same level of abstraction. We can just pass the `GitAttributesFile` to the EOL conversion type to decide the final targe EOL.
    * If the EOL feature is supported only in git backend, and the git attributes feature is available universally, we should pass `GitAttributesFile` to the abstraction that hides the implementation details of the EOL feature to decide the final EOL.
    * If the git attributes feature is supported only in git backend, but the EOL feature is available universally, we should pass the EOL configuration to the abstraction that hides the implementation details of the git attributes feature to decide the final EOL.
* How could this whole design fit the following use cases outside `local_working_copy.rs`? Will it make it more difficult to work on the following features in the future?
    * The `linguist-generated` attribute and the `jj diff` command. It should be mostly handled in `DiffRenderer`, where we could make use of `GitAttributesFile` and read from the store to fetch the git attributes associated with any diff entries.
    * The clean/smudge filter. It can be handled within `local_working_copy.rs` before a file is written to the store or to the disk. `git-crypt` can be supported.

**Comment by 06393993** [comment]
Another question from kejadlen and me:

* In https://github.com/jj-vcs/jj/compare/main...kejadlen:jj:lfs, which is a more up-to-date version of https://github.com/jj-vcs/jj/pull/6036, a new `git.ignore-filters` config replaces the original `git.ignore-lfs-files` config.
    * Are we still in favor of this approach if we commit to the clean/smudge filter? Do we think we should use this feature to close the gap for git lfs when we are working on clean/smudge filter?
    * If we are implementing this new ignore files based on `.gitattributes` filter feature, where should we store a list of filters that we want to ignore so that we can have access to the list of filters when we traversing files in `local_working_copy.rs`? https://github.com/jj-vcs/jj/compare/main...kejadlen:jj:lfs puts it in `GitAttributesFile::ignore_filters`, which I don't think is a good place, we should probably just put add a new feild to `SnapshotOptions`, `CheckoutOptions`, and `DirectoryToVisit`, and add a new parameter to `TreeState::update`. But I am wondering if the core contributor has some opinions here?

**Comment by poliorcetics** [comment]
EDIT: this is wrong, I don't know how I got it to work, retrying on a fresh repo shows it didn't actually work

~~As a workaround for people ending up here looking to ignore LFS file: recent JJ versions support [`snapshot.auto-track`](https://jj-vcs.github.io/jj/latest/config/#paths-to-automatically-track), which you can use like this:~~

```toml
# .jj/repo/config.toml

[snapshot]
auto_track = '''
  . ~ (
    root-glob:"my/first/lfs/dir/*.png"
    | root-glob:"my/second/folder/*.jpg"
  )
'''
```

**Comment by lgarron** [comment]
> As a workaround for people ending up here looking to ignore LFS file: recent JJ versions support [`snapshot.auto-track`](https://jj-vcs.github.io/jj/latest/config/#paths-to-automatically-track), which you can use like this:

Does anyone have a repro showing this working?

I'd love to use this workaround for now, but I was unable to get it working using any of a bunch of variations.

**Comment by oars** [comment]
~Somehow the original instructions from @poliorcetics are working for me at the moment on a brand new collocated clone.~

Lies. I hadn't used any of the lfs files yet which led me to believe it was working.

**Comment by mcclure** [comment]
Hello, I've been following this github issue for a while and I'm still a little confused either what the status is or what the goal is. I can't tell from this discussion if jiujitsu (1) chokes on repos with git-lfs files, (2) checks the repos out, but on git-lfs files checks out the hash placeholder files instead of the real files, or (3) checks the repo out in some acceptable state, but somehow corrupts or erases git-lfs files on next commit. Meanwhile a change is being discussed which causes jj to "ignore" git-lfs files, but I can't tell what the specific effect of this change would be either (check out hash placeholder file? omit those files from the checked out repo fully?). Could we get some clarity?

Lack of git-lfs is by itself the thing holding me back from trying jj. I work in video games where git-lfs is essentially mandatory for every project due to the omnipresence of binary assets. I currently use Mercurial which is "good enough", but the hg-git bridge cannot check out git-lfs files (with the failure mode of checking out hash placeholder files but otherwise behave properly) so if any git alternative can work with git-lfs that's a huge incentive to switch.

**Comment by 06393993** [comment]
Disclaimer: I am not the core contributor, and am not working on Jujutsu fulltime.

I plan to work on the .gitattributes clean/smudge feature, which should solve the LFS support once and for all. However I can't promise a timeline. I hope we can complete it end of this year. As you can tell from the list of design questions I posted earlier, there are too many ambiguous part that we need to nail down before implementation.

I am currently working on a design on minimal `.gitattribute` support, and how it could integrate with EOL conversion first. Will publish my design somewhere public, so that everyone knows the progress.

**Comment by mcclure** [comment]
@06393993 , thanks for explaining. Could you clarify one more thing:

> I plan to work on the .gitattributes clean/smudge feature, which should solve the LFS support once and for all.

Is the idea that supporting clean/smudge in your plan would entail the ability for the smudge step to execute external binaries, and therefore the mainline git-lfs executable could simply be invoked?

**Comment by PhilipMetzger** [comment]
> Hello, I've been following this github issue for a while and I'm still a little confused either what the status is or what the goal is. I can't tell from this discussion if jiujitsu (1) chokes on repos with git-lfs files, (2) checks the repos out, but on git-lfs files checks out the hash placeholder files instead of the real files, or (3) checks the repo out in some acceptable state, but somehow corrupts or erases git-lfs files on next commit.

Answering your questions, Jujutsu currently checks out the hash placeholder files and since it doesn't understand `.gitattributes` it doesn't do anything with it and always sees them as "modified" ( `jj previously detected these files as changed as it expected the original file content before the git lfs checkout to be present.` from #5480 ) . 

This is the _tracking issue_ following the implementation in Jujutsu, either via Gitoxide (currently unlikely since the project isn't as far yet) or external binary (like `git fetch` since Version 0.30). It's also intertwined with the issues @06393993 mentions, so there's work in progress to totally ignore Git-LFS files in a checkout (#6036 and its variants), which allows users to fallback to the `git-lfs` CLI.   

> Is the idea that supporting clean/smudge in your plan would entail the ability for the smudge step to execute external binaries, and therefore the mainline git-lfs executable could simply be invoked?

I think I answered that above.

**Comment by 06393993** [comment]
I have completed some of the most important parts of the design for a utility to query the git attributes associated with a file: https://github.com/06393993/jj/blob/gitattr-design/docs/design/gitattributes.md. Will open a PR for review once all the sections are fleshed out. If we have early comments, please don't hesitate to leave comments on the https://github.com/06393993/jj/commit/fac88b9b487546bd5d2c997b7e58a3e163b43db9 commit, or comment on the https://github.com/jj-vcs/jj/issues/80 issue or the https://github.com/jj-vcs/jj/issues/53 issue.

**Comment by adrian-brady** [comment]
I have been successfully using JJ at work with a git LFS repository. I forgot I had made my own branch with some changes else I would have updated a thread somewhere, I thought I was using kejadlen's branch.

I just tested building off  [main...adrian-brady:jj:lfs](https://github.com/jj-vcs/jj/compare/main...adrian-brady:jj:lfs) and created a colocated repo inside of a Git LFS repository and it checked out no LFS files.

I have this in my config

```toml
[git]
ignore-files = ["lfs"]
```

I do not work with the LFS files myself. Whenever I notice a problem building the app I run a `git lfs pull` and things are fine. I have not done extensive testing with edge cases either.

---

@06393993 great job with this write up

**Comment by Matthew-Benson** [comment]
I just want +1 #7098 that it is working for ignoring LFS files in my repo's. Here's how I installed and configured it if anyone less familiar with `jj` wants to test it for themselves:

```sh
cargo install --git https://github.com/gusinacio/jj.git --branch lfs --locked --bin jj jj-cli
jj config set --user git.ignore-files '["lfs"]'
jj git clone --colocate <my repo with many LFS files>
git lfs pull
git status # shows no changes
jj status  # shows no changes
```

Huge kudo's for all the hard work to get us to this point. This is the last blocker for broader adoption of `jj` at my workplace.

**Comment by lgarron** [comment]
> I just want +1 [#7098](https://github.com/jj-vcs/jj/pull/7098) that it is working for ignoring LFS files in my repo's. Here's how I installed and configured it if anyone less familiar with `jj` wants to test it for themselves:

Thanks, that first command is very helpful!

I tried it out on https://github.com/cubing/cubing.js and in case anyone else tries it on their repo I wanted to document that even with the global `git.ignore-files` config for `jj`:

- `git lfs pull` produces tracked changes for me.
- `jj abandon` seems to reset LFS files to their metadata content even if their LFS contents were properly in place.

So I'm definitely still looking forward to a full solution. But I'm really glad to see activity on it! 🤩

**Comment by 06393993** [comment]
Open the https://github.com/jj-vcs/jj/pull/7164 PR for the very first gitattributes corner stone design. Feel free to ask me to add you to the author list as long as you feel that you should be credited.

**Comment by valyagolev** [comment]
I couldn't figure out how to make it _always_ play _nice_, but for the ones who want to use it _in anger_ until this is supported, hopefully soon. This is not at all "the right way", this is "what to do when it's there and annoying and you want to go back to your work".

First of all, generally ignoring the lfs files through `git.ignore-files` works, as long as you don't touch them.

If you end up depending on a branch that changes them, or change them yourself, it's a bit of a chore, yes:

1. If your files don't have their real content but have the lfs metadata stuff: Do `git lfs pull`, you'll end up with changes in the repo according to jj. `jj commit -m "lfs stuff pls ignore me" `, and then when you're a pushing a branch, reorder commits to put this one on the very top and bookmark the one before.
2. If you must change the files, just make a `git` commit directly and jj will have to work with it.

**Comment by ghost** [comment]
Hello Team,
Do you have any ETA for this feature ?

When working with unreal engine, all the files are LFS.
It is the only thing that stops me to use JJ sadly

**Comment by Kethku** [comment]
Im also coming from gamedev and require lfs support to use jj.

**Comment by Vampire** [comment]
My work-around stab at this, at least as long as I mainly want to update non-lfs files is.

This config:
```toml
[revset-aliases]
"private()" = "..bookmarks(glob-i:'local/*')"
"private_commits()" = "private() ~ immutable() ~ ..remote_bookmarks()"
"push_bookmarks()" = "(remote_bookmarks(remote=origin)..@ ~ private_commits()) & bookmarks()"

[git]
private-commits = "private()"

[revsets]
git-push = "push_bookmarks()"
```

And then
```bash
git lfs pull
jj bookmark create local/realize-lfs-filesbook
jj new @- # creating change X
jj new @ local/realize-lfs-filesbook # creating change Y
```

Now I can do work with the LFS files realized but not changed in Y.
I can `squash` or `absorb` changes I did in Y to X.
I can `describe`, `bookmark`, and `push` X.

Not sure how it would work when needing to change lfs files though.

**Comment by jwvanderbeck** [comment]
Just chiming in as another indie game dev interested in trying out jj but can't without proper git-lfs support.

**Comment by Cidan** [comment]
Xoogler here -- for those of us without piper, this is an incredibly important feature to have. Voting on this for high priority.

**Comment by davidpdrsn** [comment]
I don't think we need to keep commenting that this feature is important to us. I'm sure the maintainers are aware. It just causes spam in peoples inboxes.

**Comment by lgarron** [comment]
> I don't think we need to keep commenting that this feature is important to us. I'm sure the maintainers are aware. It just causes spam in peoples inboxes.

As a maintainer of many open-source projects myself, I'm really sympathetic to this. But it's rather difficult to understand what's going on. Now two months since that comment:

- This issue has twice the upvotes of any other issue. So it's pretty clear that this is probably the most valuable feature that people want/need. I think it's unreasonable to expect maintainers to be at the beck and call of the community, but I don't think a maintainer has commented on this issue in months. Like, I get that an open-source project has the right to set its own priorities depending in large part on volunteer interest, but users also have the right to feel that the project is unable to prioritize important features. Over time, this leads to an awkward "can I keep using this project or do I need to migrate to a different solution?" question. I've been through that process with other projects that I've used, and it's never fun. But I would really love to stay fully invested in `jj`.
- There are two PRs that appear to be stalled but unclosed:
  - https://github.com/jj-vcs/jj/pull/6036
  - https://github.com/jj-vcs/jj/pull/7098
- There is a design doc at https://github.com/jj-vcs/jj/pull/8144 but I have no idea if it might be merged or implemented some day.

Even if I have significant experience with Rust and wanted to invest time to contribute, it's unclear what I could do without making the situation even more complicated.

Even if the situation currently is "we don't know if any sort of LFS support/ignoring can be shipped soon, please don't get invested in the idea", I would really appreciate a fresh update from the maintainers.

**Comment by 06393993** [comment]
Let me make some update in this issue then.

Git LFS support is not trivial. Essentially, it makes use of 3 gitattributes: `filter`, `merge`, and `diff`, in the order of importance. `jj` currently doesn't have any gitattributes support. Therefore, my original plan is:

- Design and implement the infrastructure to read gitattributes files
    - Design: https://github.com/jj-vcs/jj/pull/7164
    - Implementation: https://github.com/jj-vcs/jj/pull/8266, I need to carry the `text` `gitattributes` with the change, because we don't think we should merge code that is not actually used anywhere, and `text` `gitattributes` is a feature that applies conversion on check in and checkout, but doesn't introduce that many changes.
- Design and implement the basic function of the `filter` gitattributes:
    - Design: https://github.com/jj-vcs/jj/pull/8144
    - Implementation: https://github.com/jj-vcs/jj/pull/8719
    - ~~The design is still WIP,~~ It takes such long time because:
        - I am lazy(I am sorry).
        - Async Rust is [very difficult](https://www.iroh.computer/blog/async-rust-challenges-in-iroh) to be correct. We need to be async runtime agnostic, which means, we should not panic inside nor outside a tokio worker thread, and we should not block the worker thread when waiting for the child process to return. I haven't yet confirmed if we need to consider cancellation safety.
        - Error handling. git allows optional filter, which means the git command still succeeds if the filter process fails, we don't know if we want to support that or not. We need also make sure the user should be able to recover the repo to a reasonable state if the filter fails in the middle of checkout.
- File a separate issue to discuss how to convert a conflict file on snapshot and checkout. See details at https://github.com/jj-vcs/jj/pull/8266#issuecomment-3703117912 and https://github.com/06393993/jj/blob/128a297947b2a2b59ac8e62436c5a53533f5079d/docs/design/gitattributes.md#diff-conflicts-and-merge. https://github.com/jj-vcs/jj/pull/8441 is also a related PR to make this easier, ~~but it's waiting for review as well~~.
- Discuss whether we need to support the long process git `filter` feature.
- Design and implement a proper external diff tool/merge tool support. Nothing has started yet.
- Design and implement a proper `merge` `gitattributes` support. Nothing has started yet.
- Discuss whether we want to add support for the `diff` `gitattributes`, which mostly influences the internal diff tool.

It's probable that I complicate the issue and just write a bad doc which is too verbose, so I don't have enough reviewers to review my design(thanks PhilipMetzger for always timely review my PRs, but one reviewer is not enough), especially when the maintainers lack the hands to review PRs. And the maintainers are also too kind to just drive me away from the job or just completely reject my ideas. I sincerely appologize that this issue is currently in this half-broken state.

I think the maintainers want to invest in this feature, since they still occasionally discuss related topics, so it's really me who should be blamed. I am Ok if anyone thinks they are willing to replace me on this job as long as maintainers are also Ok. Until then, I will continue to slowly work on this. My next step is to clean up the gitattributes filter design doc, write an implementation, and send those 2 PRs for review.

As of the other 2 existing PRs, https://github.com/jj-vcs/jj/pull/7098 is an improved version of https://github.com/jj-vcs/jj/pull/6036. However, https://github.com/jj-vcs/jj/pull/7098 doesn't resolve all the comments, especially the ask for documentation on why files are only ignored when snapshot(check in) but not on checkout. Given the reason behind this design decision, the maintainer may or may not reject this idea.

**Comment by mering** [comment]
> Design and implement a proper external diff tool/merge tool support. Nothing has started yet.

IIUC jj already supports external diff/merge tools: https://github.com/jj-vcs/jj/blob/main/docs/config.md#editing-diffs. So this probably only needs to be wired up correctly if there is anything to do at all.

**Comment by 06393993** [comment]
> > Design and implement a proper external diff tool/merge tool support. Nothing has started yet.
> 
> IIUC jj already supports external diff/merge tools: https://github.com/jj-vcs/jj/blob/main/docs/config.md#editing-diffs. So this probably only needs to be wired up correctly if there is anything to do at all.

It's the "wire up" part that can be tricky, or too complicated that the maintainer may want to reject the feature, so we should design carefully.

**Comment by sbarfurth** [comment]
I understand this is all very complicated and I can sympathize with wanting a durable solution. At the same time I would like to acknowledge that `jj` is becoming popular with some speed (judging by GitHub stars) and people will inevitably drop it when they find it actively clashes with their existing repositories. I presume a decent chunk of such people will
never come back.

There is a "working" PR that circumvents lots of the headache by at least ignoring LFS files so they can be managed outside `jj` without complications. I personally feel there would be value in integrating something like this. This makes the experience workable and enables working on a _proper_ implementation over a longer time. Moving from ignored to properly supported might not even be a breaking change down the line.

I believe perfect is being the enemy of good here.

**Comment by 06393993** [comment]
> I understand this is all very complicated and I can sympathize with wanting a durable solution. At the same time I would like to acknowledge that `jj` is becoming popular with some speed (judging by GitHub stars) and people will inevitably drop it when they find it actively clashes with their existing repositories. I presume a decent chunk of such people will never come back.
> 
> There is a "working" PR that circumvents lots of the headache by at least ignoring LFS files so they can be managed outside `jj` without complications. I personally feel there would be value in integrating something like this. This makes the experience workable and enables working on a _proper_ implementation over a longer time. Moving from ignored to properly supported might not even be a breaking change down the line.
> 
> I believe perfect is being the enemy of good here.

My comment on #7098 is not about asking to be perfect. I believe I am just asking for more documents in the code, so that the maintainer can judge if it's an acceptable workaround. We believe it's not natural to only ignore those files on snapshot but not on update. Maintainers should also consider if #7098 is significantly a better WA than https://github.com/jj-vcs/jj/issues/80#issuecomment-3432510006, particularly considering the extra maintenance cost.

**Comment by Valodim** [comment]
Meanwhile, I have been using `github:weiznich/jj/a12505164ac655c8769dfee9991aae5ccf20990f` for a year now, and it worked just fine for me. Please do not overengineer this feature: just parse `.gitattributes` using the same mechanism as `.gitignore`, and ignore matching files during snapshotting. I don't think this is a workaround, it's just a simple way to no longer do the wrong thing (messing up lfs file) in repositories that use it. Any other use of git attributes or lfs integration is just the 2% to solving the 98% pain point of not being able to use jujutsu in lfs repos.

I know it's easy to say "just", but it really seems like we had working solutions for a big real world problem, that got held up by scope creep and YAGNI here.

**Comment by 06393993** [comment]
> Meanwhile, I have been using `github:weiznich/jj/a12505164ac655c8769dfee9991aae5ccf20990f` for a year now, and it worked just fine for me. Please do not overengineer this feature: just parse `.gitattributes` using the same mechanism as `.gitignore`, and ignore matching files during snapshotting. I don't think this is a workaround, it's just a simple way to no longer do the wrong thing (messing up lfs file) in repositories that use it. Any other use of git attributes or lfs integration is just the 2% to solving the 98% pain point of not being able to use jujutsu in lfs repos.
> 
> I know it's easy to say "just", but it really seems like we had working solutions for a big real world problem, that got held up by scope creep and YAGNI here.

Am I correct that we think we should stop investing any effort to design and implement proper support for the `filter` `gitattributes`, because it's overengineering and we are not going to need it, and just push the merge of https://github.com/jj-vcs/jj/pull/7098, because that's the real suolition. In addition, we should also revert how we read `gitattributes` files back to the state where it's similar to how we read `gitignore` files?

I am Ok getting rid of all my work and pursuit that path, but before that happens, can you check if https://github.com/06393993/jj/tree/gitattr-design-wip is really that bad and we really don't want it? Thanks.

**Comment by mering** [comment]
I don't think has to be an _either/or_ decision: Does anything speak against ignoring LFS files as an intermediate step as soon as possible while working towards full LFS support?

**Comment by jwvanderbeck** [comment]
> Am I correct that we think we should stop investing any effort to design and implement proper support for the `filter` `gitattributes`, because it's overengineering and we are not going to need it, and just push the merge of [#7098](https://github.com/jj-vcs/jj/pull/7098), because that's the real suolition. In addition, we should also revert how we read `gitattributes` files back to the state where it's similar to how we read `gitignore` files?
> 
> I am Ok getting rid of all my work and pursuit that path, but before that happens, can you check if https://github.com/06393993/jj/tree/gitattr-design-wip is really that bad and we really don't want it? Thanks.

No I understood the suggestion to be to release a workaround that allows repos to be used with `jj` while a more proper solution is worked on since it will clearly take time to implement.

**Comment by 06393993** [comment]
> > Am I correct that we think we should stop investing any effort to design and implement proper support for the `filter` `gitattributes`, because it's overengineering and we are not going to need it, and just push the merge of [#7098](https://github.com/jj-vcs/jj/pull/7098), because that's the real suolition. In addition, we should also revert how we read `gitattributes` files back to the state where it's similar to how we read `gitignore` files?
> > I am Ok getting rid of all my work and pursuit that path, but before that happens, can you check if https://github.com/06393993/jj/tree/gitattr-design-wip is really that bad and we really don't want it? Thanks.
> 
> No I understood the suggestion to be to release a workaround that allows repos to be used with `jj` while a more proper solution is worked on since it will clearly take time to implement.

Sorry for my misunderstanding, but if I call a solution overengineering, and YAGNI, I don't want that solution, because it adds to complexity and hurts maintainability. And if I call a design a real solution, I usually won't consider an immediate follow up to drastically change how the design works.

Let's wait for maintainers' opinions.

**Comment by 06393993** [comment]
I have sent the gitattr filter design for review in https://github.com/jj-vcs/jj/pull/8144, and uploaded an example implementation in https://github.com/jj-vcs/jj/pull/8719 for reference.

I have tested that #8719 works with git-lfs with the following settings:

```toml
git.filter.enabled = true

[git.filter.drivers.lfs]
required = true
clean = ["git-lfs", "clean", "--", "$path"]
smudge = ["git-lfs", "smudge", "--", "$path"]
```

Feel free to try and provide feedbacks. Thanks. If you are seeing bug, please run the command again with the `JJ_LOG=TRACE` environment variable, and attach the log. Thanks. If you are using the correct binary you should see something like the following:

```text
2026-01-27T15:08:20.138550Z TRACE run:run_internal:run_command:cmd_new:tree_state: jj_lib::filter: The filter gitattributes support is enabled.
```

EDIT: correct the setting from `working-copy.filter` to `git.filter`.

**Comment by martinvonz** [comment]
Copying from https://github.com/jj-vcs/jj/pull/7164#issuecomment-3788516060:

> We talked in the maintainer group today about how we think gitattributes should be handled. We all felt that we should only support gitattributes in the working copy. That would mean that we respect it when we update the working copy and when we snapshot the working copy. Since external diff editors are implemented using temporary working copies, we would presumably support it there too. That probably means that we should also support it in the internal diff editor (the one based on `scm-record`) for consistency.

Since LFS is based on gitattributes, that would apply to LFS too. That would mean that LFS files would be expanded in the working copy but `jj diff` would just show a diff in the hash. I hope that sounds okay.

**Comment by mcclure** [comment]
@martinvonz , the behavior you are describing is actually what `git diff` itself does with git lfs files. If it's helpful, here's an example from an LFS-using project of mine (gitattributes contains `*.jpg filter=lfs diff=lfs merge=lfs -text`).

```bash
mcc@Machine:~/projectname$ mv ~/Documents/Castlevania_Legacy_of_Darkness.jpg Tests/Resources/wikimedia-banff.jpg # Overwrite an LFS file
mcc@Machine:~/projectname$ git diff
diff --git a/Tests/Resources/wikimedia-banff.jpg b/Tests/Resources/wikimedia-banff.jpg
index c2e8149..0e9cbe1 100644
--- a/Tests/Resources/wikimedia-banff.jpg
+++ b/Tests/Resources/wikimedia-banff.jpg
@@ -1,3 +1,3 @@
 version https://git-lfs.github.com/spec/v1
-oid sha256:d8305bb374955cc893d65bc9aeb425fdb0b506c2baf5f541f25909d791a9712e
-size 28515
+oid sha256:300b680dadf7e6956861da81522dd237ea1f56e4e43aed632b9764715d283057
+size 21494
```

Moreover, LFS is *usually* used with binary files, meaning that *usually* the standard git cannot display useful diffs anyway, and I assume(?) neither can jj. So your proposed behavior is correct.

There is a second command besides diff which *is* important to support, though. This is `git lfs status`:

```bash
mcc@Machine:~/projectname$ echo "nonsense line" >> README.md  # Modify a non-LFS file
mcc@Machine:~/projectname$ git lfs status
On branch projectname-2.2.x
Objects to be pushed to origin/projectname-2.2.x


Objects to be committed:


Objects not staged for commit:

	Tests/Resources/wikimedia-banff.jpg (LFS: d8305bb -> File: 300b680)
	README.md (Git: eb33ad7 -> File: 992dd3e)
```

This command is internal to the git-lfs tool, and has the same behavior as `git status`, except it shows "the LFS view" of every altered file. This is important because it confirms which changed file are being changed "as LFS" and which are being changed "as git" (see the bit in parenthesis), i.e., it ensures no .gitattributes errors. There are multiple important `git lfs commandname` commands like this (see `git lfs help`). I assume in the new jj approach, there will be *some* way to invoke the lfs "subcommands", be it `jj lfs commandname` or somehow directly invoking lfs?

**Comment by senekor** [comment]
> There are multiple important `git lfs commandname` commands like this (see `git lfs help`). I assume in the new jj approach, there will be _some_ way to invoke the lfs "subcommands", be it `jj lfs commandname` or somehow directly invoking lfs?

`lfs` is not a built-in subcommand of git. When running `git lfs commandname`, git just searches `PATH` for executables called `git-lfs` and passes arguments through to it. jj doesn't do that intentionally, but users can either run `git-lfs comandname` directly or configure an alias, e.g.:
```
[aliases]
lfs = ["util", "exec", "--", "git-lfs"]
```

Then you can run `jj lfs commandname` with the same effect as running `git lfs commandname`.

**Comment by keysmashes** [comment]
Just a note for anybody trying to use the current version of #7098 and wondering why it's not working: the `working-copy.filter` settings table appears to have been renamed to `git.filter`.

**Comment by elephantum** [comment]
I would like to give my +1 to an incremental path to LFS in JJ.

First: stop breaking LFS files in existing repos as soon as possible

Second: do a proper thing

**Comment by martinvonz** [comment]
FWIW, I don't think any of the maintainers object to doing it incrementally. I don't think we can accept https://github.com/jj-vcs/jj/compare/main...weiznich:jj:ignore_lfs as is, however. It seems to look up all `.gitattributes` files in the repo, which will be completely unusably slow at Google (`jj status` will take hours or days to run). Or maybe it's okay for now and Google should just patch it out until we get a better solution merged into this repo. I'm curious what others think.

**Comment by SamAmco** [comment]
It might help a lot in the meantime if it were possible to have a separate jj ignore file. A file specific to jj that lets you ignore files only from jj's perspective. This way you could ignore all the lfs pointer files but still manage them with git. I don't know if that's easy though.

**Comment by martinvonz** [comment]
That wouldn't be an ignore file in the same sense as .gitignore, though, since I think you're asking for jj to also ignore changes to tracked files matching these patterns. An alternative might be to add support for negative patterns to `jj sparse`.

**Comment by SamAmco** [comment]
That sounds good. That would probably work for me in the near term and would be a useful feature to have in general.

**Comment by jgilchrist** [comment]
> Or maybe it's okay for now and Google should just patch it out until we get a better solution merged into this repo. I'm curious what others think.

Sample size of one here, but quite a few repos I contribute to use LFS and I've resorted to maintaining my own build with weiznichs' change applied as it's such a huge improvement for my workflows. I'd love to be able to introduce jj to my co-workers, but can't see them compiling a forked version to use it. So I'd personally love to see this happen.

**Comment by raylu** [comment]
I introduced it to my co-workers with a doc where step 1 was "download it from https://github.com/raylu/jj/releases"
I just enabled GH actions on my fork and created a release

**Comment by steveklabnik** [comment]
As someone who doesn't use LFS at all, if people are going so far as to distribute their own builds of jj to get this patch in, given that (in my understanding) Google already distributes their own builds of jj (for `jj piper` and whatever else), I think figuring out how to get this in and letting Google strip it out seems like a nice gesture. I have no idea how much work that would be to maintain, but Google doing the work rather than enthusiasts feels like the correct alignment of work here.

**Comment by martinvonz** [comment]
> Google doing the work rather than enthusiasts feels like the correct alignment of work here.

I agree. That shouldn't be a problem. My concern was whether https://github.com/jj-vcs/jj/compare/main...weiznich:jj:ignore_lfs is good enough for the project in general. I haven't tested it myself in a large repo. I suspect it's fine.

**Comment by martinvonz** [comment]
To set expectations, I don't know anyone who plans to work on a short-term solution. I think @06393993 is only working on the full support for LFS (actually writing the large files to the working copy). If others want it sooner, consider sending a PR. Some version of @weiznich's commit may be enough, but please test it in a large repo (the Linux repo and the nixpkgs repos are common test cases) and share the performance numbers in the commit description.

If someone wants to teach `jj sparse` to support filesets, I think that would also be appreciated. It should be sufficient for it to support a single fileset, but how to add and remove patterns with `jj sparse set` is not obvious from UX point of view.

**Comment by weiznich** [comment]
Given that my changes are still discussed here and the main blocker now seems to be performance I would like to clarify one rather important thing: My change gates the behaviour behind a off-by-default setting, so while I can understand that it might be non-optimal in terms of performance it just won't affect the performance for existing repositories. So anyone using this on an repo without lfs files won't be affected by this at all, as they can just not change the default setting. I do not know anything about the google intern usage, but I suspect that this won't be affected at all due to the disabled setting.
It might have problems if a user enables this setting for a really large repository that contains LFS files. I think it's important to note: This is a case that currently also doesn't work (due to missing lfs support), so it's not like this is breaking anything that previously worked.
(Well there is also the case that a user enables this setting, without having LFS files. In this case I would just dismiss this as configuration issue)

**Comment by martinvonz** [comment]
I was thinking of proposing that we gate it behind a config. Good to hear that that's already done. Maybe someone (you?) can just take that commit and send it for review then? Has that already been attempted and we rejected it? Sorry if that was the case :(

**Comment by weiznich** [comment]
There is https://github.com/jj-vcs/jj/pull/7098 which contains another more recent version of that feature. It is waiting for a decision for more than half a year now. Maybe it's a good idea to move that PR forward instead?

**Comment by steveklabnik** [comment]
@martinvonz yeah that's absolutely fair; I want jj to remain viable in large repos too, I was just trying to think through the social side of it only, you're right to call out that there's technical aspects of "should we merge this" too.

I see since I last read the thread, the "put it behind a config" bit has already been suggested, so, I have nothing more to say, haha. I do hope we can get something in to all parties' satisfaction here, sounds like there's a good path forward.

**Comment by PhilipMetzger** [comment]
> There is [#7098](https://github.com/jj-vcs/jj/pull/7098) which contains another more recent version of that feature. It is waiting for a decision for more than half a year now. Maybe it's a good idea to move that PR forward instead?

I agree as someone who spent quite a bit of time on both @06393993's version and it to _get_ any progress. It just lacked maintainer review to bring it over the line.

**Comment by mikelu92** [comment]
hi just wanted to add that I appreciate everyone's work here and I'm hoping that whatever solution that's worked out here would also work for [git-crypt](https://github.com/AGWA/git-crypt) as I have repos that use it. I know it's a bit different than git-lfs but both use .gitattributes 

currently using a version of #7098 and it is working for me on my own fork to ignore these encrypted files

**Comment by martinvonz** [comment]
FYI, #7098 was closed by the author, but @kejadlen offered to pick up the remaining work.

**Comment by jgilchrist** [comment]
For anybody who might've been following this issue, but not following PRs, @kejadlen's PR is up: https://github.com/jj-vcs/jj/pull/9068

**Comment by temportalflux** [comment]
Very excited to see some progress on LFS support. I'll be keeping a close eye out for changes and hoping jj will one day be a competent alternative for usage in game dev spaces. There's a related fairly dormant discussion/post over here (looking for consistency moreso than anything else) https://github.com/jj-vcs/jj/discussions/4306

**Comment by gusinacio** [comment]
In case anyone *really* needs support to Git Lfs, I added full support for it with the help of AI here: https://github.com/gusinacio/jj/commits/git-lfs-support.

Edit: I don't think this is the right way to solve and that's why I'm not pushing a PR for this. My projects were getting clunky with git lfs and jj and I decided to share it as a short term solution until git lfs support is added.

Most of us that really need Lfs were already using some form of outdated fork anyway.

**Comment by julian7** [comment]
I'd really love to see a solution for this, to finish 1.0 for my [redact](julian7/redact) project. Direct compatibility with jj and git would be awesome, but I wouldn't mind if there was a different solution added to jj eventually.

**Comment by steeleduncan** [comment]
> In case anyone _really_ needs support to Git Lfs, I added full support for it with the help of AI here: https://github.com/gusinacio/jj/commits/git-lfs-support.
> 
> Edit: I don't think this is the right way to solve and that's why I'm not pushing a PR for this. My projects were getting clunky with git lfs and jj and I decided to share it as a short term solution until git lfs support is added.
> 
> Most of us that really need Lfs were already using some form of outdated fork anyway.

@gusinacio Could I ask why this is the wrong approach? I wouldn't mind using lfs with jj, so I'm wondering if it is "wrong but works", or "wrong, will cause serious issues in edge cases"?

**Comment by gusinacio** [comment]
> > In case anyone _really_ needs support to Git Lfs, I added full support for it with the help of AI here: https://github.com/gusinacio/jj/commits/git-lfs-support.
> > 
> > Edit: I don't think this is the right way to solve and that's why I'm not pushing a PR for this. My projects were getting clunky with git lfs and jj and I decided to share it as a short term solution until git lfs support is added.
> > 
> > Most of us that really need Lfs were already using some form of outdated fork anyway.
> 
> @gusinacio Could I ask why this is the wrong approach? I wouldn't mind using lfs with jj, so I'm wondering if it is "wrong but works", or "wrong, will cause serious issues in edge cases"?



I looked briefly at the code, but I'm no jj maintainer and I don't want to take credit for what the AI did and I barely reviewed. The world is moving to a direction where open source maintainers are being flooded with PRs and Issues fully generated by AI and if I was one of the maintainers I wouldn't want some AI slop that a random guy just vibed to be inside the project I take care of.

The code itself works, I did some QA with all my projects, fixed some issues with performance and all of that.

The "wrong" part is that I can't be responsible for the code the AI did and I barely reviewed.

I'm sharing this in case anyone needs a band-aid solution, and also in case anyone wants to "study" and create the right way (where you know 100% of the changes you are making).

If I made a PR, I would be a man in the middle prompting to the AI the review. I'd rather let the reviewers do this by themselves.

**Comment by Draqun** [comment]
Hi. I'm also waiting for Git LFS support, as it's currently the only blocker preventing me from using jj in several projects. Leaving a comment here so I can follow the progress — and to show that this feature matters to users.

Thanks for all the hard work on jj. I'm cheering for the project and looking forward to seeing how it evolves.
Best regard.

**Comment by trcoffman** [comment]
This would be so amazing if it were added!

**Comment by yetisunny** [comment]
This would be great indeed

**Comment by LukeCarrier** [comment]
A reminder that you do not need to leave comments on issues in order to subscribe. The notifications are in the sidebar over there 👉 and you can leave a 👍 on the issue up there ☝️ to signal to maintainers that you too are interested in this feature.

Apologies for spamming everybody already subscribed to this issue with yet another unhelpful comment, for four in a row. I will accept my rotten tomato throwing gracefully.
